drizzlepac/generate_final_product_filenames.py

```diff
@@ -1,15 +1,160 @@
+    obset_table = Table.read(results, format='ascii.fast_no_header')
+    # label each column with a descriptive name
+    for col, cname in zip(obset_table.colnames, colnames):
+        obset_table[col].name = cname
+    # Add INSTRUMENT column
+    instr = INSTRUMENT_DICT[obset_table['filename'][0][0]]
+    # convert input to an Astropy Table for parsing
+    obset_table.add_column(Column([instr]*len(obset_table)), name='instrument')
+    # parse Table into a tree-like dict
+    obset_tree = build_obset_tree(obset_table)
+    # Now create final dict
+    obset_dict = parse_obset_tree(obset_tree)
+
+    return obset_dict
+
+# Translate the database query on an obset into actionable lists of filenames
+def build_obset_tree(obset_table):
+    """Convert obset table into a tree listing all products to be created."""
+
+    # Each product will consist of the appropriate string as the key and
+    # a dict of 'info' and 'files' information
+
+    # Start interpreting the obset table
+    obset_tree = {}
+    for row in obset_table:
+        # Get some basic information from the first row - no need to check
+        # for multiple instruments as data from different instruments will
+        # not be combined.
+        det = row['detector']
+        filt = row['filters']
+        row_info, filename = create_row_info(row)
+        # Initial population of the obset tree for this detector
+        if det not in obset_tree:
+            obset_tree[det] = {}
+            obset_tree[det][filt] = [(row_info, filename)]
+        else:
+            det_node = obset_tree[det]
+            if filt not in det_node:
+                det_node[filt] = [(row_info, filename)]
+            else:
+                filt_node = det_node[filt]
+                filt_node.append((row_info, filename))
+
+    return obset_tree
+
+def create_row_info(row):
+    """Build info string for a row from the obset table"""
+    info_list = [str(row['proposal_id']), row['program_id'], row['instrument'],
+                 row['detector'], row['filters'], row['filename'][:row['filename'].find('_')]]
+    return ' '.join(map(str.upper, info_list)), row['filename']
+
+def parse_obset_tree(det_tree):
+    """Convert tree into products
+
+    Tree generated by `create_row_info()` will always have the following
+    levels:
+          * Detector
+              * filters
+                  * exposure
+    Each exposure will have an entry dict with keys 'info' and 'filename'.
+
+    Products created will be:
+      * total detection product
+      * filter products
+      * single exposure product
+    """
+    # Initialize products dict
+    obset_products = {}
+
+    # for each level, define a product, starting with the detector used...
+    sep_indx = 0
+    tdp = TDP_STR.format(0)
+    obset_products[tdp] = {'info': "", 'files': []}
+
+    # setup products for each detector used
+    for det in det_tree.keys():
+        filt_tree = det_tree[det]
```

Comment:
Iterate over dictionary values directly instead of keys(): `for filt_tree in det_tree.values():`

Reply:
Done.